luminoth/models/fasterrcnn/rcnn_target.py

```diff
@@ -76,97 +68,164 @@ def _build(self, proposals, gt_boxes):
-            disable_inds = np.random.choice(
-                fg_inds, size=(len(fg_inds) - num_fg), replace=False
-            )
-            # We disable them with their negatives just to be able to debug
-            # down the road.
-            proposals_label[disable_inds] = - proposals_label[disable_inds]
+        best_fg_labels_for_proposals = tf.add(
+            tf.gather(gt_boxes[:, 4], overlaps_best_gt_idxs),
+            1.
+        )
+        iou_is_fg = tf.greater_equal(
+            max_overlaps, self._foreground_threshold
+        )
+        best_proposals_idxs = tf.argmax(overlaps, axis=0)
+
+        # Set the indices in best_proposals_idxs to True, and the rest to
+        # false.
+        # tf.sparse_to_dense is used because we know the set of indices which
+        # we want to set to True, and we know the rest of the indices
+        # should be set to False. That's exactly the use case of
+        # tf.sparse_to_dense.
+        is_best_box = tf.sparse_to_dense(
+            sparse_indices=tf.reshape(best_proposals_idxs, [-1]),
+            sparse_values=True, default_value=False,
+            output_shape=tf.cast(proposals_label_shape, tf.int64),
+            validate_indices=False
+        )
 
-        if len(fg_inds) < num_fg:
-            # When our foreground samples are not as much as we would like them
-            # to be, we log a warning message.
-            tf.logging.warning(
-                'We\'ve got only {} foreground samples instead of {}.'.format(
-                    len(fg_inds), num_fg
-                )
+        fg_condition = tf.logical_or(
+            iou_is_fg, is_best_box
+        )
+        # We update proposals_label with the value in
+        # best_fg_labels_for_proposals only when the box is foreground, or it
+        # is the best proposal for a certain gt_box.
+        proposals_label = tf.where(
+            condition=fg_condition,
+            x=best_fg_labels_for_proposals,
+            y=proposals_label
+        )
+
+        # proposals_label now has a value in [0, num_classes + 1] for
+        # proposals we are going to use and -1 for the ones we should ignore.
+        # But we still need to make sure we don't have a number of proposals
+        # higher than minibatch_size * foreground_fraction.
+        max_fg = int(self._foreground_fraction * self._minibatch_size)
+        fg_inds = tf.where(
+            condition=fg_condition
+        )
+
+        def disable_some_fgs():
+            # We want to delete a randomly-selected subset of fg_inds of
+            # size `fg_inds.shape[0] - max_fg`.
+            # We shuffle along the dimension 0 and then we get the first
+            # num_fg_inds - max_fg indices and we disable them.
+            if self._debug:
+                shuffled_inds = tf.random_shuffle(fg_inds, seed=0)
+            else:
+                shuffled_inds = tf.random_shuffle(fg_inds)
+            disable_place = (tf.shape(fg_inds)[0] - max_fg)
```

Comment:
It should be the case that `num_fg_inds > max_fg` because `disable_some_fgs` only runs if the condition `tf.greater(tf.shape(fg_inds)[0], max_fg)` evaluates to true.

It may warrant adding
```
with tf.control_dependencies([tf.assert_non_negative(disable_inds, message="something")]):
    disable_inds = shuffled_inds[:disable_place]
```